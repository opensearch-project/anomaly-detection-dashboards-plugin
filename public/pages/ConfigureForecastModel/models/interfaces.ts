/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

// Formik values used when creating the model configuration
export interface ModelConfigurationFormikValues {
  shingleSize: number;
  imputationOption?: ImputationFormikValues;
  interval: number | undefined;
  windowDelay: number | undefined;
  suggestedSeasonality?: number;
  recencyEmphasis?: number;
  resultIndexMinAge?: number | string;
  resultIndexMinSize?: number | string;
  resultIndexTtl?:number | string;
  flattenCustomResultIndex?: boolean;
  resultIndex?: string;
  horizon?: number;
  history?: number;
}

export interface ImputationFormikValues {
  imputationMethod?: string;
  custom_value?: CustomValueFormikValues[];
}

export interface CustomValueFormikValues {
  // forecasting has only one feature and we don't need to specify the feature name
  featureName?: string;
  data: number;
}
