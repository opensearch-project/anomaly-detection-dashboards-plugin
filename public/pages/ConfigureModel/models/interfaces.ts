/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import { AggregationOption } from '../../../models/types';

// Formik values used when creating/editing the model configuration
export interface ModelConfigurationFormikValues {
  featureList: FeaturesFormikValues[];
  categoryFieldEnabled: boolean;
  categoryField: string[];
  shingleSize: number;
  imputationOption?: ImputationFormikValues;
  suppressionRules?: RuleFormikValues[];
}

export interface FeaturesFormikValues {
  featureId: string;
  featureName: string | undefined;
  featureType: FEATURE_TYPE;
  featureEnabled: boolean;
  aggregationQuery: string;
  aggregationBy?: string;
  aggregationOf?: AggregationOption[];
  newFeature?: boolean;
}

export interface ImputationFormikValues {
  imputationMethod?: string;
  custom_value?: CustomValueFormikValues[];
}

export interface CustomValueFormikValues {
  featureName: string;
  data: number;
}

export interface RuleFormikValues {
  featureName: string;
  absoluteThreshold?: number;
  relativeThreshold?: number;
  aboveBelow: string;
}
