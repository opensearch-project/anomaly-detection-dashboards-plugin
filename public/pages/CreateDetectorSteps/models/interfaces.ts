/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { DetectorDefinitionFormikValues } from '../../DefineDetector/models/interfaces';
import { ModelConfigurationFormikValues } from '../../ConfigureModel/models/interfaces';
import { DetectorJobsFormikValues } from '../../DetectorJobs/models/interfaces';

// Formik values used upon creation (includes all fields + those related to historical detector date range)
export interface CreateDetectorFormikValues
  extends DetectorDefinitionFormikValues,
    ModelConfigurationFormikValues,
    DetectorJobsFormikValues {}
