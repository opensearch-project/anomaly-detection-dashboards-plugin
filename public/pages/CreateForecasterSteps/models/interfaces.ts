/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { ForecasterDefinitionFormikValues } from '../../DefineForecaster/models/interfaces';
import { ModelConfigurationFormikValues } from '../../ConfigureForecastModel/models/interfaces';

// Formik values used upon creation (includes all fields + those related to historical detector date range)
export interface CreateForecasterFormikValues
  extends ForecasterDefinitionFormikValues,
    ModelConfigurationFormikValues {}
