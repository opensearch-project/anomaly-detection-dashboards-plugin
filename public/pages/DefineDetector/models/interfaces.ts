/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { UIFilter } from '../../../models/interfaces';

// Formik values used when creating/editing the detector definition
export interface DetectorDefinitionFormikValues {
  name: string;
  description: string;
  index: { label: string }[];
  resultIndex?: string;
  filters: UIFilter[];
  filterQuery: string;
  timeField: string;
  interval: number;
  windowDelay: number;
  resultIndexMinAge?: number;
  resultIndexMinSize?: number;
  resultIndexTtl?:number;
}
