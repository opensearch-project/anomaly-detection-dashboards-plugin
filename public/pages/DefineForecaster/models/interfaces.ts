/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { UIFilter } from '../../../models/interfaces';
import { FEATURE_TYPE } from '../../../models/interfaces';
import { AggregationOption } from '../../../models/types';
import { ClusterOption } from '../utils/helpers';

// Formik values used when creating/editing the forecaster definition
export interface ForecasterDefinitionFormikValues {
  name: string;
  description: string;
  index: { label: string }[];
  filters: UIFilter[];
  filterQuery: string;
  timeField: string;
  clusters?: ClusterOption[]; 
  featureList: FeaturesFormikValues[];
  categoryFieldEnabled: boolean;
  categoryField: string[];
}

export interface FeaturesFormikValues {
  featureId: string;
  featureName: string | undefined;
  featureType: FEATURE_TYPE;
  featureEnabled: boolean;
  aggregationQuery: string;
  aggregationBy?: string;
  aggregationOf?: AggregationOption[];
  newFeature?: boolean;
}
