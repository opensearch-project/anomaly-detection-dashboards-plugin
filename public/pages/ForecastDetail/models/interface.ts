/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { UIFilter } from "public/models/interfaces";
import { ImputationFormikValues } from "../../ConfigureForecastModel/models/interfaces";
import { FeaturesFormikValues } from "../../DefineForecaster/models/interfaces";
import { ClusterOption } from "../../DefineForecaster/utils/helpers";

// Formik values used when editing the model configuration
export interface DetailsFormikValues {
    name: string;
    description: string;
    resultIndex?: string;
    resultIndexMinAge?: number | string;
    resultIndexMinSize?: number | string;
    resultIndexTtl?: number | string;
    flattenCustomResultIndex?: boolean;
}

export interface ConfigureFormikValues {
    index: { label: string }[];
    filters: UIFilter[];
    filterQuery: string;
    timeField: string;
    clusters?: ClusterOption[];
    featureList: FeaturesFormikValues[];
    categoryFieldEnabled: boolean;
    categoryField: string[];
    shingleSize: number;
    imputationOption?: ImputationFormikValues;
    interval: number | undefined;
    windowDelay: number | undefined;
    suggestedSeasonality?: number;
    recencyEmphasis?: number;
    horizon?: number;
    history?: number;
}
