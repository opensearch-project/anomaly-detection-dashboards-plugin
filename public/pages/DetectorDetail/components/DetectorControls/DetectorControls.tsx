/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { useState } from 'react';
import {
  EuiButton,
  EuiContextMenuItem,
  EuiContextMenuPanel,
  EuiFlexGroup,
  EuiFlexItem,
  EuiPopover,
} from '@elastic/eui';
import { Detector } from '../../../../models/interfaces';

interface DetectorControls {
  onEditDetector(): void;
  onEditFeatures(): void;
  onDelete(): void;
  onStartDetector(): void;
  onStopDetector(): void;
  detector: Detector;
}
export const DetectorControls = (props: DetectorControls) => {
  const [isOpen, setIsOpen] = useState(false);
  return (
    <EuiFlexGroup justifyContent="spaceBetween" alignItems="center">
      <EuiFlexItem grow={false}>
        <EuiPopover
          id="actionsPopover"
          button={
            <EuiButton
              iconType="arrowDown"
              iconSide="right"
              data-test-subj="actionsButton"
              onClick={() => setIsOpen(!isOpen)}
            >
              Actions
            </EuiButton>
          }
          panelPaddingSize="none"
          anchorPosition="downLeft"
          isOpen={isOpen}
          closePopover={() => setIsOpen(false)}
        >
          <EuiContextMenuPanel>
            <EuiContextMenuItem
              key="editDetector"
              data-test-subj="editDetector"
              onClick={props.onEditDetector}
            >
              Edit detector settings
            </EuiContextMenuItem>

            <EuiContextMenuItem
              key="editFeatures"
              data-test-subj="editFeature"
              onClick={props.onEditFeatures}
            >
              Edit model configuration
            </EuiContextMenuItem>

            <EuiContextMenuItem
              key="deleteDetector"
              data-test-subj="deleteDetector"
              onClick={props.onDelete}
              style={{ color: '#FF6666' }}
            >
              Delete detector
            </EuiContextMenuItem>
          </EuiContextMenuPanel>
        </EuiPopover>
      </EuiFlexItem>
    </EuiFlexGroup>
  );
};
