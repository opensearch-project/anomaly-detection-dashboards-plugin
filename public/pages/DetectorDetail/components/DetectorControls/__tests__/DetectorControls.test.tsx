/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { DetectorControls } from '../DetectorControls';
import { UNITS } from '../../../../../models/interfaces';

describe('<DetectorControls /> spec', () => {
  const detector = {
    primaryTerm: 1,
    seqNo: 1,
    id: '',
    name: '4QY4YHEB5W9C7vlb3Mou',
    description: 'test detector',
    timeField: 'timestamp',
    indices: ['test-index'],
    filterQuery: { match_all: { boost: 1.0 } },
    featureAttributes: [],
    windowDelay: { period: { interval: 0, unit: UNITS.MINUTES } },
    detectionInterval: { period: { interval: 0, unit: UNITS.MINUTES } },
    uiMetadata: {},
    lastUpdateTime: 1586714882000,
    enabled: false,
    enabledTime: 1586722082000,
  };
  const features = [
    {
      featureId: 'ZNU5YHEBTQyq-K19pyEP',
      featureName: 'feature1',
      featureEnabled: true,
      importance: 1,
      aggregationQuery: {
        feature1: { sum: { field: 'value' } },
      },
    },
  ];
  const onEditDetector = jest.fn();
  const onEditFeatures = jest.fn();
  const onDelete = jest.fn();
  const onStartDetector = jest.fn();
  const onStopDetector = jest.fn();
  describe('Detector Controls', () => {
    test('renders detector controls with disabled detecor and empty features', () => {
      const { container } = render(
        <DetectorControls
          onEditDetector={onEditDetector}
          onEditFeatures={onEditFeatures}
          onDelete={onDelete}
          onStartDetector={onStartDetector}
          onStopDetector={onStopDetector}
          // @ts-ignore
          detector={detector}
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders detector controls with enabled detecor and empty features', () => {
      const enabledDetector = { ...detector, enabled: true };
      const { container } = render(
        <DetectorControls
          onEditDetector={onEditDetector}
          onEditFeatures={onEditFeatures}
          onDelete={onDelete}
          onStartDetector={onStartDetector}
          onStopDetector={onStopDetector}
          // @ts-ignore
          detector={enabledDetector}
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders detector controls with disabled detecor and features', () => {
      const disabledDetectorWithFeatures = {
        ...detector,
        featureAttributes: features,
      };
      const { container } = render(
        <DetectorControls
          onEditDetector={onEditDetector}
          onEditFeatures={onEditFeatures}
          onDelete={onDelete}
          onStartDetector={onStartDetector}
          onStopDetector={onStopDetector}
          // @ts-ignore
          detector={disabledDetectorWithFeatures}
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders detector controls with enabled detecor and features', () => {
      const disabledDetectorWithFeatures = {
        ...detector,
        enabled: true,
        featureAttributes: features,
      };
      const { container } = render(
        <DetectorControls
          onEditDetector={onEditDetector}
          onEditFeatures={onEditFeatures}
          onDelete={onDelete}
          onStartDetector={onStartDetector}
          onStopDetector={onStopDetector}
          // @ts-ignore
          detector={disabledDetectorWithFeatures}
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
