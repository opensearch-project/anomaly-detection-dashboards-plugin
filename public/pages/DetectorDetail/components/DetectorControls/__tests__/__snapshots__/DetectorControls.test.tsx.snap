// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DetectorControls /> spec Detector Controls renders detector controls with disabled detecor and empty features 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft"
        id="actionsPopover"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButton euiButton--primary"
            data-test-subj="actionsButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButton__content"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              />
              <span
                class="euiButton__text"
              >
                Actions
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<DetectorControls /> spec Detector Controls renders detector controls with disabled detecor and features 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft"
        id="actionsPopover"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButton euiButton--primary"
            data-test-subj="actionsButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButton__content"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiButtonContent__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                  fill-rule="non-zero"
                />
              </svg>
              <span
                class="euiButton__text"
              >
                Actions
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<DetectorControls /> spec Detector Controls renders detector controls with enabled detecor and empty features 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft"
        id="actionsPopover"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButton euiButton--primary"
            data-test-subj="actionsButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButton__content"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiButtonContent__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                  fill-rule="non-zero"
                />
              </svg>
              <span
                class="euiButton__text"
              >
                Actions
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<DetectorControls /> spec Detector Controls renders detector controls with enabled detecor and features 1`] = `
<div>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      class="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <div
        class="euiPopover euiPopover--anchorDownLeft"
        id="actionsPopover"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButton euiButton--primary"
            data-test-subj="actionsButton"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButton__content"
            >
              <svg
                aria-hidden="true"
                class="euiIcon euiIcon--medium euiButtonContent__icon"
                focusable="false"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                  fill-rule="non-zero"
                />
              </svg>
              <span
                class="euiButton__text"
              >
                Actions
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
